airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/poetry/publish/context.py

```diff
@@ -0,0 +1,95 @@
+#
+# Copyright (c) 2023 Airbyte, Inc., all rights reserved.
+#
+
+import os
+from datetime import datetime
+from typing import Optional
+
+from pipelines.airbyte_ci.connectors.context import PipelineContext
+from pipelines.airbyte_ci.connectors.publish.context import PublishConnectorContext
+
+
+class PyPIPublishContext(PipelineContext):
+    def __init__(
+        self,
+        pypi_token: str,
+        package_path: str,
+        report_output_prefix: str,
+        is_local: bool,
+        git_branch: str,
+        git_revision: str,
+        ci_report_bucket: Optional[str] = None,
+        registry: Optional[str] = None,
+        gha_workflow_run_url: Optional[str] = None,
+        dagger_logs_url: Optional[str] = None,
+        pipeline_start_timestamp: Optional[int] = None,
+        ci_context: Optional[str] = None,
+        ci_gcs_credentials: Optional[str] = None,
+        package_name: Optional[str] = None,
+        version: Optional[str] = None,
+    ) -> None:
+        self.pypi_token = pypi_token
+        self.registry = registry or "https://pypi.org/simple"
+        self.package_path = package_path
+        self.package_name = package_name
+        self.version = version
+
+        pipeline_name = f"Publish PyPI {package_path}"
+
+        super().__init__(
+            pipeline_name=pipeline_name,
+            report_output_prefix=report_output_prefix,
+            ci_report_bucket=ci_report_bucket,
+            is_local=is_local,
+            git_branch=git_branch,
+            git_revision=git_revision,
+            gha_workflow_run_url=gha_workflow_run_url,
+            dagger_logs_url=dagger_logs_url,
+            pipeline_start_timestamp=pipeline_start_timestamp,
+            ci_context=ci_context,
+            ci_gcs_credentials=ci_gcs_credentials,
+        )
+
+    @staticmethod
+    async def from_publish_connector_context(connector_context: PublishConnectorContext) -> Optional["PyPIPublishContext"]:
+        """
+        Create a PyPIPublishContext from a ConnectorContext.
+
+        The metadata of the connector is read from the current workdir to capture changes that are not yet published.
+        If pypi is not enabled, this will return None.
+        """
+
+        current_metadata = connector_context.connector.metadata
+        if (
+            "remoteRegistries" not in current_metadata
+            or "pypi" not in current_metadata["remoteRegistries"]
+            or not current_metadata["remoteRegistries"]["pypi"]["enabled"]
+        ):
+            return None
+
+        version = current_metadata["dockerImageTag"]
+        if connector_context.pre_release:
+            # use current date as pre-release version
+            rc_tag = datetime.now().strftime("%Y%m%d%H%M")
```

Comment:
```suggestion
            release_candidate_tag = datetime.now().strftime("%Y%m%d%H%M")
```
Can we somehow use the [same tagging logic as for docker pre-release images](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/publish/context.py#L93) using commit sha? Would be cool to use a common function to tag pre-release for docker images and pypi packages. And I kinda like the fact that tagging / version is handled at the Context level.

Reply:
We could do the git commit hash thing when using local version identifiers: https://packaging.python.org/en/latest/specifications/version-specifiers/#local-version-identifiers

But they need to be prefixed with a `+`, so it won't be 100% the same. Still a bit better than the current local logic, I will adjust.

Reply:
Just tested and unfortunately it doesn't work - test.pypi is rejecting local versions:
> 400 '0.2.1+391b68cd52' is an invalid value for Version. Error:   
         Can't use PEP 440 local versions. See                                  
         https://packaging.python.org/specifications/core-metadata for more     
         information.

Public version identifiers need to be numeric, I will revert back to the old logic.